File: backend/services/websocket-tts-service.ts
/**
 * WebSocket TTS Service
 * Handles real-time text-to-speech streaming using Socket.io and OpenAI TTS API
 */

import { Server as SocketIOServer, Socket } from 'socket.io';
import { Server as HTTPServer } from 'http';

interface TTSRequest {
  text: string;
  voice?: 'alloy' | 'echo' | 'fable' | 'onyx' | 'nova' | 'shimmer';
  speed?: number;
}

interface ConnectionInfo {
  socket: Socket;
  isStreaming: boolean;
  connectedAt: Date;
}

export class WebSocketTTSService {
  private io: SocketIOServer;
  private connections: Map<string, ConnectionInfo> = new Map();
  private apiKey: string;

  constructor(httpServer: HTTPServer, apiKey: string, corsOrigins: string[]) {
    this.apiKey = apiKey;

    // Initialize Socket.io with CORS configuration
    this.io = new SocketIOServer(httpServer, {
      cors: {
        origin: corsOrigins,
        credentials: true,
      },
      transports: ['websocket', 'polling'],
    });

    this.setupEventHandlers();
    console.log('[TTS Service] WebSocket TTS service initialized');
    console.log('[TTS Service] Socket.io listening on:', corsOrigins);
  }

  /**
   * Setup Socket.io event handlers
   */
  private setupEventHandlers(): void {
    console.log('[TTS Service] Setting up event handlers');

    this.io.on('connection', (socket: Socket) => {
      console.log(`[TTS Service] Client connected: ${socket.id}`);

      // Store connection info
      this.connections.set(socket.id, {
        socket,
        isStreaming: false,
        connectedAt: new Date(),
      });

      // Handle speech generation request
      socket.on('generate-speech', async (data: TTSRequest, callback) => {
        try {
          await this.handleSpeechGeneration(socket, data);
          if (callback) callback({ success: true });
        } catch (error) {
          console.error('[TTS Service] Error generating speech:', error);
          socket.emit('tts-error', {
            message: error instanceof Error ? error.message : 'Failed to generate speech',
          });
          if (callback) callback({ success: false, error: 'Failed to generate speech' });
        }
      });

      // Handle stop speech request
      socket.on('stop-speech', () => {
        const conn = this.connections.get(socket.id);
        if (conn) {
          conn.isStreaming = false;
          console.log(`[TTS Service] Speech stopped for client: ${socket.id}`);
        }
      });

      // Handle disconnection
      socket.on('disconnect', () => {
        console.log(`[TTS Service] Client disconnected: ${socket.id}`);
        this.connections.delete(socket.id);
      });

      // Send connection confirmation
      console.log(`[TTS Service] Sending tts-connected event to ${socket.id}`);
      socket.emit('tts-connected', {
        message: 'TTS service ready',
        socketId: socket.id,
      });
      console.log(`[TTS Service] tts-connected event sent to ${socket.id}`);
    });
  }

  /**
   * Handle speech generation request
   */
  private async handleSpeechGeneration(socket: Socket, data: TTSRequest): Promise<void> {
    const conn = this.connections.get(socket.id);
    if (!conn) {
      throw new Error('Connection not found');
    }

    // Prevent concurrent streaming for the same client
    if (conn.isStreaming) {
      console.log(`[TTS Service] Client ${socket.id} is already streaming, ignoring request`);
      return;
    }

    // Validate input
    if (!data.text || data.text.trim().length === 0) {
      throw new Error('Text is required');
    }

    if (data.text.length > 4000) {
      throw new Error('Text is too long (max 4000 characters)');
    }

    conn.isStreaming = true;

    try {
      console.log(`[TTS Service] Generating speech for client ${socket.id}: "${data.text.substring(0, 50)}..."`);

      // Emit start event
      socket.emit('speech-start', {
        text: data.text,
        voice: data.voice || 'alloy',
      });

      // Call OpenAI TTS API
      const response = await fetch('https://api.openai.com/v1/audio/speech', {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.apiKey}`,
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          model: 'tts-1', // Use tts-1 for lower latency, tts-1-hd for higher quality
          input: data.text,
          voice: data.voice || 'alloy',
          speed: data.speed || 1.0,
          response_format: 'mp3',
        }),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`OpenAI TTS API error: ${response.status} - ${errorText}`);
      }

      // Stream the audio data
      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error('Failed to get response reader');
      }

      let totalBytes = 0;
      const chunkSize = 16384; // 16KB chunks for optimal streaming

      while (true) {
        // Check if client stopped or disconnected
        if (!conn.isStreaming || !this.connections.has(socket.id)) {
          console.log(`[TTS Service] Streaming interrupted for client ${socket.id}`);
          reader.cancel();
          break;
        }

        const { done, value } = await reader.read();

        if (done) {
          break;
        }

        totalBytes += value.length;

        // Emit audio chunk
        socket.emit('audio-chunk', {
          data: Buffer.from(value).toString('base64'),
          index: Math.floor(totalBytes / chunkSize),
        });

        // Small delay to prevent overwhelming the client
        await new Promise(resolve => setTimeout(resolve, 5));
      }

      // Emit end event
      socket.emit('speech-end', {
        totalBytes,
        duration: totalBytes / 16000, // Approximate duration
      });

      console.log(`[TTS Service] Speech generation completed for client ${socket.id} (${totalBytes} bytes)`);

    } catch (error) {
      console.error('[TTS Service] Error in speech generation:', error);
      socket.emit('tts-error', {
        message: error instanceof Error ? error.message : 'Unknown error',
      });
    } finally {
      conn.isStreaming = false;
    }
  }

  /**
   * Get service statistics
   */
  public getStats() {
    return {
      activeConnections: this.connections.size,
      connections: Array.from(this.connections.entries()).map(([id, conn]) => ({
        socketId: id,
        isStreaming: conn.isStreaming,
        connectedAt: conn.connectedAt,
      })),
    };
  }

  /**
   * Shutdown the service
   */
  public shutdown(): void {
    console.log('[TTS Service] Shutting down...');
    this.connections.clear();
    this.io.close();
  }
}

File: backend-python/app/services/websocket_tts_service.py
"""
WebSocket TTS Service
Handles real-time text-to-speech streaming using Socket.io and OpenAI TTS API
"""

import asyncio
import base64
import logging
from dataclasses import dataclass, field
from datetime import datetime
from typing import Any

import httpx
import socketio

from app.config import get_settings

logger = logging.getLogger(__name__)


@dataclass
class ConnectionInfo:
    """Connection information for a client."""
    sid: str
    is_streaming: bool = False
    connected_at: datetime = field(default_factory=datetime.utcnow)


class WebSocketTTSService:
    """WebSocket TTS Service for real-time text-to-speech streaming."""

    def __init__(self, cors_origins: list[str] | None = None):
        self.settings = get_settings()
        self.connections: dict[str, ConnectionInfo] = {}

        # Initialize Socket.io async server
        self.sio = socketio.AsyncServer(
            async_mode="asgi",
            cors_allowed_origins=cors_origins or ["*"],
            logger=False,
            engineio_logger=False,
        )

        self._setup_event_handlers()
        logger.info("[TTS Service] WebSocket TTS service initialized")
        logger.info(f"[TTS Service] CORS allowed origins: {cors_origins or ['*']}")

    def _setup_event_handlers(self) -> None:
        """Setup Socket.io event handlers."""
        logger.info("[TTS Service] Setting up event handlers")

        @self.sio.event
        async def connect(sid: str, environ: dict[str, Any]) -> None:
            logger.info(f"[TTS Service] Client connected: {sid}")
            self.connections[sid] = ConnectionInfo(sid=sid)

            # Send connection confirmation
            logger.info(f"[TTS Service] Sending tts-connected event to {sid}")
            await self.sio.emit(
                "tts-connected",
                {"message": "TTS service ready", "socketId": sid},
                to=sid,
            )
            logger.info(f"[TTS Service] tts-connected event sent to {sid}")

        @self.sio.event
        async def disconnect(sid: str) -> None:
            logger.info(f"[TTS Service] Client disconnected: {sid}")
            self.connections.pop(sid, None)

        @self.sio.on("generate-speech")
        async def on_generate_speech(sid: str, data: dict[str, Any]) -> dict[str, Any]:
            try:
                await self._handle_speech_generation(sid, data)
                return {"success": True}
            except Exception as e:
                logger.error(f"[TTS Service] Error generating speech: {e}")
                await self.sio.emit(
                    "tts-error",
                    {"message": str(e)},
                    to=sid,
                )
                return {"success": False, "error": "Failed to generate speech"}

        @self.sio.on("stop-speech")
        async def on_stop_speech(sid: str) -> None:
            conn = self.connections.get(sid)
            if conn:
                conn.is_streaming = False
                logger.info(f"[TTS Service] Speech stopped for client: {sid}")

    async def _handle_speech_generation(self, sid: str, data: dict[str, Any]) -> None:
        """Handle speech generation request."""
        conn = self.connections.get(sid)
        if not conn:
            raise ValueError("Connection not found")

        # Prevent concurrent streaming for the same client
        if conn.is_streaming:
            logger.info(f"[TTS Service] Client {sid} is already streaming, ignoring request")
            return

        # Validate input
        text = data.get("text", "")
        if not text or not text.strip():
            raise ValueError("Text is required")

        if len(text) > 4000:
            raise ValueError("Text is too long (max 4000 characters)")

        conn.is_streaming = True

        try:
            voice = data.get("voice", "alloy")
            speed = data.get("speed", 1.0)

            logger.info(f"[TTS Service] Generating speech for client {sid}: \"{text[:50]}...\"")

            # Emit start event
            await self.sio.emit(
                "speech-start",
                {"text": text, "voice": voice},
                to=sid,
            )

            # Call OpenAI TTS API
            api_key = self.settings.openai_api_key
            if not api_key:
                raise ValueError("OpenAI API key not configured")

            async with httpx.AsyncClient() as client:
                response = await client.post(
                    "https://api.openai.com/v1/audio/speech",
                    headers={
                        "Authorization": f"Bearer {api_key}",
                        "Content-Type": "application/json",
                    },
                    json={
                        "model": "tts-1",  # Use tts-1 for lower latency, tts-1-hd for higher quality
                        "input": text,
                        "voice": voice,
                        "speed": speed,
                        "response_format": "mp3",
                    },
                    timeout=60.0,
                )

                if response.status_code != 200:
                    raise ValueError(f"OpenAI TTS API error: {response.status_code} - {response.text}")

                # Stream the audio data
                total_bytes = 0
                chunk_size = 16384  # 16KB chunks for optimal streaming

                async for chunk in response.aiter_bytes(chunk_size):
                    # Check if client stopped or disconnected
                    if not conn.is_streaming or sid not in self.connections:
                        logger.info(f"[TTS Service] Streaming interrupted for client {sid}")
                        break

                    total_bytes += len(chunk)

                    # Emit audio chunk
                    await self.sio.emit(
                        "audio-chunk",
                        {
                            "data": base64.b64encode(chunk).decode("utf-8"),
                            "index": total_bytes // chunk_size,
                        },
                        to=sid,
                    )

                    # Small delay to prevent overwhelming the client
                    await asyncio.sleep(0.005)

                # Emit end event
                await self.sio.emit(
                    "speech-end",
                    {
                        "totalBytes": total_bytes,
                        "duration": total_bytes / 16000,  # Approximate duration
                    },
                    to=sid,
                )

                logger.info(f"[TTS Service] Speech generation completed for client {sid} ({total_bytes} bytes)")

        except Exception as e:
            logger.error(f"[TTS Service] Error in speech generation: {e}")
            await self.sio.emit(
                "tts-error",
                {"message": str(e)},
                to=sid,
            )
        finally:
            conn.is_streaming = False

    def get_stats(self) -> dict[str, Any]:
        """Get service statistics."""
        return {
            "activeConnections": len(self.connections),
            "connections": [
                {
                    "socketId": sid,
                    "isStreaming": conn.is_streaming,
                    "connectedAt": conn.connected_at.isoformat(),
                }
                for sid, conn in self.connections.items()
            ],
        }

    def get_asgi_app(self) -> socketio.ASGIApp:
        """Get the ASGI app for mounting in FastAPI."""
        return socketio.ASGIApp(self.sio)

    async def shutdown(self) -> None:
        """Shutdown the service."""
        logger.info("[TTS Service] Shutting down...")
        self.connections.clear()


# Singleton instance (will be initialized in main.py)
tts_service: WebSocketTTSService | None = None


def get_tts_service() -> WebSocketTTSService | None:
    """Get the TTS service instance."""
    return tts_service


def init_tts_service(cors_origins: list[str] | None = None) -> WebSocketTTSService:
    """Initialize the TTS service."""
    global tts_service
    tts_service = WebSocketTTSService(cors_origins)
    return tts_service


def get_socket_app(fastapi_app: Any) -> socketio.ASGIApp:
    """
    Create a Socket.io ASGI app that wraps the FastAPI app.
    This allows Socket.io to handle /socket.io/ requests and pass
    everything else to FastAPI.
    """
    if tts_service is None:
        raise RuntimeError("TTS service not initialized. Call init_tts_service first.")
    return socketio.ASGIApp(tts_service.sio, other_asgi_app=fastapi_app)
